bootstrap/kubeadm/controllers/kubeadmconfig_controller_test.go

```diff
@@ -194,6 +195,52 @@ func TestKubeadmConfigReconciler_Reconcile_ReturnEarlyIfMachineHasBootstrapData(
 	}
 }
 
+func TestKubeadmConfigReconciler_Reconcile_MigrateToSecret(t *testing.T) {
+	cluster := newCluster("cluster")
+	cluster.Status.InfrastructureReady = true
+	machine := newMachine(cluster, "machine")
+	config := newKubeadmConfig(machine, "cfg")
+	config.Status.Ready = true
+	config.Status.BootstrapData = []byte("test")
+	objects := []runtime.Object{
+		cluster,
+		machine,
+		config,
+	}
+	myclient := fake.NewFakeClientWithScheme(setupScheme(), objects...)
+
+	k := &KubeadmConfigReconciler{
+		Log:    log.Log,
+		Client: myclient,
+	}
+
+	request := ctrl.Request{
+		NamespacedName: types.NamespacedName{
+			Namespace: "default",
+			Name:      "cfg",
+		},
+	}
+
+	result, err := k.Reconcile(request)
+	if err != nil {
+		t.Fatalf("failed to reconcile:\n %+v", err)
+	}
+	if result.Requeue == true {
+		t.Fatal("did not expect to requeue")
+	}
+	if result.RequeueAfter != time.Duration(0) {
+		t.Fatal("did not expect to requeue after")
+	}
+
+	if err := k.Client.Get(context.TODO(), types.NamespacedName{Name: config.Name, Namespace: config.Namespace}, config); err != nil {
+		t.Fatalf("failed to get KubeadmConfig:\n %+v", err)
+	}
+
+	if config.Status.DataSecretName == nil {
```

Comment:
Do we also want to test that `config.Status.BootstrapData == nil` after reconciling?

Reply:
I had that initially, but it seems there is some issue "patching" nil with the fake client, need to test it live at some point

Reply:
More info here https://kubernetes.slack.com/archives/C8TSNPY4T/p1575577403348500